tests/interfaces/test_batch_torch_qnode.py

```diff
@@ -0,0 +1,1099 @@
+        res1 = circ(x1)
+        res1.backward(torch.Tensor([1, 1]))
+
+        assert np.allclose(x1.grad, expected_grad(x1))
+        assert circ.device.num_executions == 1
+        spy.assert_called_with(mocker.ANY, use_device_state=mocker.ANY)
+
+    def test_resuse_state_multiple_evals(self, mocker, tol):
+        """Tests that the Torch interface reuses the device state for adjoint differentiation,
+        even where there are intermediate evaluations."""
+        dev = qml.device("default.qubit", wires=2)
+
+        x_val = 0.543
+        y_val = -0.654
+        x = torch.tensor(x_val, requires_grad=True)
+        y = torch.tensor(y_val, requires_grad=True)
+
+        @qnode(dev, diff_method="adjoint", interface="torch")
+        def circuit(x, y):
+            qml.RX(x, wires=[0])
+            qml.RY(y, wires=[1])
+            qml.CNOT(wires=[0, 1])
+            return qml.expval(qml.PauliZ(0))
+
+        spy = mocker.spy(dev, "adjoint_jacobian")
+
+        res1 = circuit(x, y)
+        assert np.allclose(res1.detach(), np.cos(x_val), atol=tol, rtol=0)
+
+        # intermediate evaluation with different values
+        res2 = circuit(torch.tan(x), torch.cosh(y))
+
+        # the adjoint method will continue to compute the correct derivative
+        res1.backward()
+        assert np.allclose(x.grad.detach(), -np.sin(x_val), atol=tol, rtol=0)
+        assert dev.num_executions == 2
+        spy.assert_called_with(mocker.ANY, use_device_state=mocker.ANY)
+
+
+@pytest.mark.parametrize("dev_name,diff_method,mode", qubit_device_and_diff_method)
+class TestQubitIntegration:
+    """Tests that ensure various qubit circuits integrate correctly"""
+
+    def test_probability_differentiation(self, dev_name, diff_method, mode, tol):
+        """Tests correct output shape and evaluation for a tape
+        with prob and expval outputs"""
+
+        if diff_method == "adjoint":
+            pytest.skip("The adjoint method does not currently support returning probabilities")
+
+        dev = qml.device(dev_name, wires=2)
+        x_val = 0.543
+        y_val = -0.654
+        x = torch.tensor(x_val, requires_grad=True)
+        y = torch.tensor(y_val, requires_grad=True)
+
+        @qnode(dev, diff_method=diff_method, mode=mode, interface="torch")
+        def circuit(x, y):
+            qml.RX(x, wires=[0])
+            qml.RY(y, wires=[1])
+            qml.CNOT(wires=[0, 1])
+            return qml.probs(wires=[0]), qml.probs(wires=[1])
+
+        res = circuit(x, y)
+
+        expected = np.array(
+            [
+                [np.cos(x_val / 2) ** 2, np.sin(x_val / 2) ** 2],
+                [
+                    (1 + np.cos(x_val) * np.cos(y_val)) / 2,
+                    (1 - np.cos(x_val) * np.cos(y_val)) / 2,
+                ],
+            ]
+        )
+
+        if diff_method == "backprop":
+            # TODO: check why this differs from other interfaces
+            expected = expected.flatten()

```

Comment:
Wait, `expected` is just a `np.array`, and the parameters it's based on are floats, why would there be any difference?